### PR TITLE
Add continuation functionality

### PR DESCRIPTION
**Pull request by dudnyk**
Add Continuation template classes and implementation
which provide functionality to create a chain of tasks,
and execute tasks.
Add continuation unit tests.

Relates-To: OLPEDGE-2077

Signed-off-by: Yevhenii Dudnyk <ext-yevhenii.dudnyk@here.com>

### CODE REVIEW

File: olp-cpp-sdk-core/src/thread/Continuation.cpp [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+      // go to the next task in the queue to finilize the task
+      context_copy->tasks_.pop();
+      return false;
+    }
+
+    auto callback = [this, context_copy, &current_task](void* arg) mutable {
+      // limit lifetime of the context shared_ptr to the function scope, for
+      // extra safety in case user's async function stores callback somewhere
+      // for forever
+      auto context = std::move(context_copy);
```
Comment (andescu) [inline olp-cpp-sdk-core/src/thread/Continuation.cpp:69]: Capturing something in the lambda, even if you think you move it actually you only move a copy of it. Are you sure this actually leaves the lambda captured variable empty?

File: olp-cpp-sdk-core/src/thread/Continuation.cpp [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+    auto scheduler = task_scheduler_.lock();
+    if (scheduler) {
+      scheduler->ScheduleTask(std::move(current_task_queue_element));
+    }
+  }
+
+ private:
+  struct ProcessorInternal {
+    std::weak_ptr<olp::thread::TaskScheduler> task_scheduler_;
```
Comment (andescu) [inline olp-cpp-sdk-core/src/thread/Continuation.cpp:137]: Why does it need to be weak_ptr. Can it survive the TaskContinuation/Continuation?

File: olp-cpp-sdk-core/src/thread/Continuation.cpp [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+                        std::pair<ContinuationImpl::AsyncTaskType,
+                                  ContinuationImpl::TaskType>& task) {
+    auto& current_task = task;
+    if (!current_task.first) {
+      // go to the next task in the queue to finilize the task
+      context_copy->tasks_.pop();
+      return false;
+    }
+
+    auto callback = [this, context_copy, &current_task](void* arg) mutable {
```
[the remark this answers is not in the document]
Comment (dudnyk) [reply olp-cpp-sdk-core/src/thread/Continuation.cpp]: done

File: olp-cpp-sdk-core/src/thread/Continuation.cpp (hunk 2 above)
Comment (dudnyk) [reply olp-cpp-sdk-core/src/thread/Continuation.cpp]: removed TaskScheduler from Processor and ProcessorInternal.

File: olp-cpp-sdk-core/include/olp/core/thread/Continuation.inl [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  if (impl_.Cancelled()) {
+    finally_callback_(client::ApiError::Cancelled());
+    finally_callback_ = nullptr;
+    return;
+  }
+
+  impl_.Run([this](void* input, bool cancelled) {
+    if (cancelled) {
+      finally_callback_(client::ApiError::Cancelled());
+    } else {
```
[the remark this answers is not in the document]
Comment (dudnyk) [reply olp-cpp-sdk-core/include/olp/core/thread/Continuation.inl:47]: done in FinishQueueExecution

File: olp-cpp-sdk-core/include/olp/core/thread/Continuation.h [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+  void Clear();
+
+ private:
+  std::shared_ptr<thread::TaskScheduler> task_scheduler_;
+  std::deque<ContinuationTask> tasks_;
+  ExecutionContext execution_context_;
+
+  /// a variable shows whether it is allow to do the changes to the task chain,
+  /// used to prevent using methods of `Continuation` after it was run.
+  bool continuation_change_allowed{true};
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/thread/Continuation.h:134]: Or during it runs. Can the user simply call Run() on the continuation and then call Then() without you crashing?

```suggestion
  bool change_allowed_{true};
```

File: olp-cpp-sdk-core/src/thread/Continuation.cpp [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+  }
+
+  void FinishQueueExecution(const std::shared_ptr<ProcessorInternal>& context) {
+    if (context->final_callback_) {
+      // the final execution context, all tasks are done or the execution
+      // is cancelled
+      context->final_callback_(context->LastOutput(), context->IsCancelled());
+      while (!context->tasks_.empty()) {
+        context->tasks_.pop();
+      }
```
Comment (andescu) [inline olp-cpp-sdk-core/src/thread/Continuation.cpp:90]: Is this not the same as calling `context->tasks_.clear()`

File: olp-cpp-sdk-core/include/olp/core/thread/Continuation.h (hunk 5 above)
Comment (dudnyk) [reply olp-cpp-sdk-core/include/olp/core/thread/Continuation.h:134]: renamed the variable.
Added a unit test which checks that Then, Finally and Run don't cause an exceptions
and aren't called if methods were already called once.

File: olp-cpp-sdk-core/src/thread/Continuation.cpp (hunk 6 above)
Comment (dudnyk) [reply olp-cpp-sdk-core/src/thread/Continuation.cpp:90]: replaced queue with deque to be able to clear a container in a simple way.

**Review by andescu — APPROVED** [review]

**Review by jmoro3 — APPROVED** [review]

**Pull request merged**